lib/file_store/s3_store.rb

```diff
@@ -105,6 +104,23 @@ def path_for(upload)
       FileStore::LocalStore.new.path_for(upload) if url && url[/^\/[^\/]/]
     end
 
+    def url_for(upload)
+      if upload.private?
+        if Rails.configuration.multisite
```

Comment:
This code is repeated below, so it would likely be better as a `get_upload_key(upload)` function that could be re-used.